File: project_rustio/src/facilities.rs
// facilities.rs
// Types for representing production buildings and building classes from DSP

use std::fmt;

use num::rational::Rational64;

use crate::rationalutilities::prettystring;


//////////////////////////////////////////
// FacilityCategory Type Implementation //
//////////////////////////////////////////

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct FacilityCategory {
	name: String,
}

impl FacilityCategory {
	pub fn new(name: &str) -> Self {
		let cleanedname = name.trim();
		if cleanedname.is_empty() {
			panic!("facility category name empty or all whitespace");
		}
		Self { name: cleanedname.to_owned() }
	}

	pub fn name(&self) -> &str {
		&self.name
	}
}

impl fmt::Display for FacilityCategory {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}", self.name())
	}
}


//////////////////////////////////
// Facility Type Implementation //
//////////////////////////////////

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Facility {
	category: FacilityCategory,
	adjective: String,
	speed: Rational64,
}

impl Facility {
	pub fn new(
		category: FacilityCategory,
		adjective: String,
		speed: Rational64
	) -> Self {
		Self { category, adjective, speed }
	}

	pub fn with_categoryname(
		categoryname: &str,
		adjective: String,
		speed: Rational64
	) -> Self {
		let category = FacilityCategory::new(categoryname);
		Self { category, adjective, speed }
	}

	pub fn category(&self) -> &FacilityCategory {
		&self.category
	}

	pub fn categoryname(&self) -> &str {
		&self.category.name()
	}
	
	pub fn speed(&self) -> Rational64 {
		self.speed
	}

    pub fn name(&self) -> String {
		if self.adjective.is_empty() {
			format!("{}", self.category())
		} else {
			format!("{} {}", &self.adjective, self.category())
		}
	}
}

impl fmt::Display for Facility {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{} ({}x)", self.name(), prettystring(self.speed()))
	}
}

File: julia_code/facilities.jl
# facilities.jl
# Types for representing production buildings and building classes from DSP


########################################
# FacilityCategory Type Implementation #
########################################

struct FacilityCategory
	name::String
	
	function FacilityCategory(name::AbstractString)
		cleanedname = strip(name)
		isempty(cleanedname) &&
			error("facility category name empty or all whitespace")
		new(string(cleanedname))
	end
end

name(fc::FacilityCategory) = fc.name

Base.show(io::IO, fc::FacilityCategory) = print(io, name(fc))


################################
# Facility Type Implementation #
################################

struct Facility
	category::FacilityCategory
	adjective::String
	speed::Rational{Int}
end

function Facility(categoryname::AbstractString, adjective::AbstractString,
	speed::Real)
	
	Facility(FacilityCategory(categoryname), adjective, speed)
end

category(f::Facility) = f.category
speed(f::Facility) = f.speed

function name(f::Facility)
	if isempty(f.adjective)
		return string(f.category)
	else
		return string(f.adjective, " ", f.category)
	end
end

function Base.show(io::IO, f::Facility)
	print(io, name(f))
	print(io, " (")
	print(io, prettystring(speed(f)))
	print(io, "x)")
end

